sudoku.py

```diff
@@ -0,0 +1,34 @@
+def is_valid(num, row, col, puzzle):
+    if num in puzzle[row]:
+        return False
+
+    for i in range(9):
+        if puzzle[i][col] == num:
+            return False
+
+    rowstart = 3 * (row // 3)
+    colstart = 3 * (col // 3)
+
+    for r in range(rowstart, rowstart + 3):
+        for c in range(colstart, colstart + 3):
+            if puzzle[r][c] == num:
+                return False
+
+    return True
+
+def solve(puzzle):
+    for row in range(9):
+        for col in range(9):
+            if puzzle[row][col] == 0:
+                for num in range(1, 10):
+                    if is_valid(num, row, col, puzzle):
+                        puzzle[row][col] = num
+                        if solve(puzzle):
```

Comment:
Nice recursive solution. In my opinion this is the easiest way to solve this kind of problem.

In what case can you _not_ use recursion?

Reply:
I know there are cases where recursion may not be the best option, however in terms of completely being unable to use it, I can not think of any.

Reply:
```
def find_max_depth(n=0):
    print(f"Depth: {n}")
    return find_max_depth(n + 1)
```